// @dyingc/brave-search-mcp-server/dist/utils/apiKeyManager.js
/**
 * API Key Manager for multi-key support with quota tracking
 * Designed for free tier keys (2000 requests/month per key)
 */
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as os from 'node:os';
// Free tier constants
const FREE_TIER_MONTHLY_QUOTA = 2000;
const DEFAULT_PER_SECOND_LIMIT = 1;
// State file path
const STATE_FILE = path.join(os.homedir(), '.brave-mcp', 'state.json');
class ApiKeyManager {
    keys;
    selectionStrategy;
    temperature;
    explorationRate; // Probability of random exploration
    constructor(apiKeys, options = {}) {
        this.keys = new Map();
        const savedStates = this.loadState();
        const { initialQuota = FREE_TIER_MONTHLY_QUOTA, selectionStrategy = 'soft-max', temperature = 1.0, explorationRate = 0.1, // Default 10% exploration rate
         } = options;
        this.selectionStrategy = selectionStrategy;
        this.temperature = temperature;
        this.explorationRate = explorationRate;
        for (const key of apiKeys) {
            const keyHash = key.slice(0, 8);
            const savedRemaining = savedStates.get(keyHash);
            this.keys.set(key, {
                key,
                remainingQuota: savedRemaining ?? initialQuota,
                monthLimit: FREE_TIER_MONTHLY_QUOTA,
                remainingSecond: DEFAULT_PER_SECOND_LIMIT,
                secondLimit: DEFAULT_PER_SECOND_LIMIT,
                isValid: true,
                inUse: false,
            });
        }
    }
    /**
     * Load saved state from file, filtering for current month only
     * Returns Map of keyHash -> remaining quota
     */
    loadState() {
        try {
            if (!fs.existsSync(STATE_FILE)) {
                return new Map();
            }
            const data = JSON.parse(fs.readFileSync(STATE_FILE, 'utf-8'));
            const now = new Date();
            const result = new Map();
            for (const [keyHash, info] of Object.entries(data.keys)) {
                const timestamp = new Date(info.timestamp);
                // Only use state from current month
                if (timestamp.getMonth() === now.getMonth() &&
                    timestamp.getFullYear() === now.getFullYear()) {
                    result.set(keyHash, info.remaining);
                }
            }
            return result;
        }
        catch (err) {
            console.warn(`[ApiKeyManager] Failed to load state: ${err.message}`);
            return new Map();
        }
    }
    /**
     * Save current state to file
     */
    saveState() {
        try {
            const dir = path.dirname(STATE_FILE);
            if (!fs.existsSync(dir)) {
                fs.mkdirSync(dir, { recursive: true });
            }
            const data = {
                version: 1,
                keys: {},
            };
            for (const [key, state] of this.keys.entries()) {
                const keyHash = key.slice(0, 8);
                data.keys[keyHash] = {
                    remaining: state.remainingQuota,
                    timestamp: new Date().toISOString(),
                };
            }
            fs.writeFileSync(STATE_FILE, JSON.stringify(data, null, 2));
        }
        catch (err) {
            console.warn(`[ApiKeyManager] Failed to save state: ${err.message}`);
        }
    }
    /**
     * Select the best available key based on remaining quota
     * Uses epsilon-greedy strategy: explore randomly with probability epsilon,
     * otherwise exploit using the configured selection strategy
     * Returns null if no valid keys available
     */
    selectBestKey() {
        // Filter valid keys with remaining quota and not in use
        const availableKeys = Array.from(this.keys.values()).filter((k) => k.isValid && k.remainingQuota > 0 && !k.inUse);
        // No available keys
        if (availableKeys.length === 0) {
            return null;
        }
        // Single available key - return directly
        if (availableKeys.length === 1) {
            availableKeys[0].inUse = true;
            return availableKeys[0].key;
        }
        let selected;
        // Epsilon-greedy exploration: random selection with probability epsilon
        if (Math.random() < this.explorationRate) {
            selected = availableKeys[Math.floor(Math.random() * availableKeys.length)];
        }
        else {
            switch (this.selectionStrategy) {
                case 'greedy-min':
                    selected = this.selectGreedyMin(availableKeys);
                    break;
                case 'soft-max':
                    selected = this.selectKeyWithSoftMax(availableKeys, this.temperature);
                    break;
                case 'greedy-max':
                default:
                    selected = this.selectGreedyMax(availableKeys);
                    break;
            }
        }
        // selected should never be null here since we checked availableKeys.length >= 2
        if (!selected) {
            return null;
        }
        selected.inUse = true;
        return selected.key;
    }
    /**
     * Select key with maximum remaining quota (greedy-max)
     */
    selectGreedyMax(keys) {
        keys.sort((a, b) => b.remainingQuota - a.remainingQuota);
        const maxQuota = keys[0].remainingQuota;
        const topKeys = keys.filter((k) => k.remainingQuota === maxQuota);
        return topKeys.length > 1 ? topKeys[Math.floor(Math.random() * topKeys.length)] : topKeys[0];
    }
    /**
     * Select key with minimum remaining quota (greedy-min)
     */
    selectGreedyMin(keys) {
        keys.sort((a, b) => a.remainingQuota - b.remainingQuota);
        const minQuota = keys[0].remainingQuota;
        const bottomKeys = keys.filter((k) => k.remainingQuota === minQuota);
        return bottomKeys.length > 1
            ? bottomKeys[Math.floor(Math.random() * bottomKeys.length)]
            : bottomKeys[0];
    }
    /**
     * Calculate soft max probabilities for key selection
     * Uses standard soft max formula: softmax_i(z;T) = exp(z_i/T) / sum(exp(z_j/T))
     * where z_i = -remainingQuota / 2000 (lower remaining = higher selection probability)
     *
     * Normalization by 2000 (FREE_TIER_MONTHLY_QUOTA) keeps scores in reasonable range [-1, 0]
     *
     * Example with T=1:
     * - Key with 1 remaining: score = -1/2000 ≈ -0.0005, exp ≈ 0.9995
     * - Key with 10 remaining: score = -10/2000 ≈ -0.005, exp ≈ 0.995
     * - Key with 1000 remaining: score = -1000/2000 = -0.5, exp ≈ 0.607
     * - Lower remaining quota yields higher selection probability
     *
     * Probabilities always sum to 1 by definition of softmax formula
     */
    calculateSoftMaxProbabilities(keys, temperature) {
        // Normalize by 2000 and negate: lower remaining = higher score
        const scores = keys.map((k) => -k.remainingQuota / FREE_TIER_MONTHLY_QUOTA);
        // Standard soft max: scale by temperature
        const scaledScores = scores.map((s) => s / temperature);
        // Numerical stability: subtract max value
        const maxScore = Math.max(...scaledScores);
        const expScores = scaledScores.map((s) => Math.exp(s - maxScore));
        const sumExp = expScores.reduce((sum, exp) => sum + exp, 0);
        const probabilities = expScores.map((exp) => exp / sumExp);
        return keys.map((key, index) => ({
            key,
            probability: probabilities[index],
        }));
    }
    /**
     * Select key using soft max probability distribution
     * temperature = 0: hard limit (equivalent to greedy-min)
     */
    selectKeyWithSoftMax(keys, temperature) {
        // Empty array
        if (keys.length === 0) {
            return null;
        }
        // temperature = 0: hard limit, equivalent to greedy-min
        if (temperature === 0) {
            return this.selectGreedyMin(keys);
        }
        const keyProbabilities = this.calculateSoftMaxProbabilities(keys, temperature);
        // Weighted random sampling (roulette wheel selection)
        const random = Math.random();
        let cumulativeProbability = 0;
        for (const { key, probability } of keyProbabilities) {
            cumulativeProbability += probability;
            if (random <= cumulativeProbability) {
                return key;
            }
        }
        // Fallback to last key
        return keyProbabilities[keyProbabilities.length - 1].key;
    }
    /**
     * Release a key after request completion
     */
    releaseKey(apiKey) {
        const state = this.keys.get(apiKey);
        if (state) {
            state.inUse = false;
        }
    }
    /**
     * Parse rate limit header: "0, 1950" -> { perSecond: 0, perMonth: 1950 }
     */
    parseRateLimitHeader(header) {
        if (!header) {
            return { perSecond: DEFAULT_PER_SECOND_LIMIT, perMonth: FREE_TIER_MONTHLY_QUOTA };
        }
        const parts = header.split(',').map((v) => parseInt(v.trim()));
        return {
            perSecond: parts[0] ?? DEFAULT_PER_SECOND_LIMIT,
            perMonth: parts[1] ?? FREE_TIER_MONTHLY_QUOTA,
        };
    }
    /**
     * Update key state from response headers
     */
    updateKeyFromHeaders(apiKey, headers) {
        const state = this.keys.get(apiKey);
        if (!state)
            return;
        const remaining = this.parseRateLimitHeader(headers.get('X-RateLimit-Remaining'));
        const limit = this.parseRateLimitHeader(headers.get('X-RateLimit-Limit'));
        state.remainingSecond = remaining.perSecond;
        state.remainingQuota = remaining.perMonth;
        state.monthLimit = limit.perMonth;
        state.secondLimit = limit.perSecond;
        state.inUse = false;
        // Save state after update
        this.saveState();
    }
    /**
     * Mark a key as invalid (e.g., 401 Unauthorized)
     */
    markKeyInvalid(apiKey, reason) {
        const state = this.keys.get(apiKey);
        if (state) {
            state.isValid = false;
            state.inUse = false;
            console.warn(`[ApiKeyManager] Key ${apiKey.slice(0, 8)}... marked invalid: ${reason}`);
        }
    }
    /**
     * Mark a key as rate-limited (429)
     */
    markKeyRateLimited(apiKey, retryAfter) {
        const state = this.keys.get(apiKey);
        if (state) {
            state.remainingQuota = 0;
            state.inUse = false;
            console.warn(`[ApiKeyManager] Key ${apiKey.slice(0, 8)}... rate-limited${retryAfter ? ` (retry after ${retryAfter}ms)` : ''}`);
        }
    }
    /**
     * Get status of all keys for debugging
     */
    getStatus() {
        return Array.from(this.keys.values()).map((k) => ({
            key: k.key.slice(0, 8) + '...',
            remaining: k.remainingQuota,
            limit: k.monthLimit,
            isValid: k.isValid,
        }));
    }
}
// Singleton instance
let apiKeyManagerInstance = null;
export function initApiKeyManager(apiKeys, options) {
    apiKeyManagerInstance = new ApiKeyManager(apiKeys, options);
}
export function getApiKeyManager() {
    return apiKeyManagerInstance;
}
