// @dyingc/brave-search-mcp-server/dist/utils/retry.js
import { RETRY_CONFIG } from '../constants.js';
/**
 * Calculate delay with full jitter to prevent thundering herd problem
 * @param baseDelay - The base delay before applying jitter
 * @returns Delay in milliseconds with jitter applied
 */
export function calculateDelayWithJitter(baseDelay) {
    // Full jitter: random value between 0 and baseDelay
    return Math.random() * baseDelay;
}
/**
 * Calculate exponential backoff delay for a given attempt
 * @param attempt - The attempt number (0-based)
 * @param baseDelay - Base delay in milliseconds
 * @param maxDelay - Maximum delay in milliseconds
 * @param multiplier - Exponential backoff multiplier
 * @returns Delay in milliseconds
 */
export function calculateExponentialDelay(attempt, baseDelay, maxDelay, multiplier) {
    // Calculate exponential delay: baseDelay * (multiplier ^ attempt)
    const exponentialDelay = baseDelay * Math.pow(multiplier, attempt);
    // Cap at maxDelay
    return Math.min(exponentialDelay, maxDelay);
}
/**
 * Check if an error is retryable based on its status code or type
 * @param error - The error to check
 * @param retryableStatusCodes - Array of retryable HTTP status codes
 * @returns True if the error should trigger a retry
 */
export function isRetryableError(error, retryableStatusCodes) {
    // Check if it's a BraveApiError with a retryable status code
    if (error instanceof BraveApiError) {
        return retryableStatusCodes.includes(error.statusCode);
    }
    // Check if it's a network error (TypeError from fetch)
    if (error instanceof TypeError) {
        const errorMsg = error.message.toLowerCase();
        return (errorMsg.includes('fetch') ||
            errorMsg.includes('network') ||
            errorMsg.includes('econnreset') ||
            errorMsg.includes('etimedout') ||
            errorMsg.includes('enotfound'));
    }
    return false;
}
/**
 * Sleep for a specified duration
 * @param ms - Duration to sleep in milliseconds
 * @returns Promise that resolves after the specified duration
 */
export function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
/**
 * Execute a function with automatic retry on failure
 * Uses exponential backoff with full jitter to prevent thundering herd
 *
 * @param fn - The function to execute (should return a Promise)
 * @param options - Retry configuration options
 * @returns Promise that resolves with the result or rejects after max retries
 *
 * @example
 * ```typescript
 * const result = await withRetry(
 *   () => fetch('https://api.example.com/data'),
 *   { maxRetries: 3, baseDelay: 1000 }
 * );
 * ```
 */
export async function withRetry(fn, options = {}) {
    const config = {
        maxRetries: options.maxRetries ?? RETRY_CONFIG.MAX_RETRIES,
        baseDelay: options.baseDelay ?? RETRY_CONFIG.BASE_DELAY_MS,
        maxDelay: options.maxDelay ?? RETRY_CONFIG.MAX_DELAY_MS,
        backoffMultiplier: options.backoffMultiplier ?? RETRY_CONFIG.BACKOFF_MULTIPLIER,
        enableJitter: options.enableJitter ?? RETRY_CONFIG.ENABLE_JITTER,
        retryableStatusCodes: options.retryableStatusCodes ?? RETRY_CONFIG.RETRYABLE_STATUS_CODES,
        onRetry: options.onRetry,
    };
    let lastError;
    const startTime = Date.now();
    for (let attempt = 0; attempt <= config.maxRetries; attempt++) {
        try {
            // First attempt (attempt 0) executes immediately
            if (attempt > 0) {
                // Calculate delay for this attempt
                const baseDelay = calculateExponentialDelay(attempt - 1, config.baseDelay, config.maxDelay, config.backoffMultiplier);
                // Apply jitter if enabled
                const delay = config.enableJitter ? calculateDelayWithJitter(baseDelay) : baseDelay;
                // Notify before retrying
                config.onRetry?.(attempt, lastError, delay);
                // Wait before retrying
                await sleep(delay);
            }
            // Execute the function
            const result = await fn();
            // Log successful recovery if we retried
            if (attempt > 0) {
                const totalDuration = Date.now() - startTime;
                console.info(`Request succeeded on attempt ${attempt} after ${totalDuration}ms`);
            }
            return result;
        }
        catch (error) {
            lastError = error;
            // Check if we should retry
            const shouldRetry = attempt < config.maxRetries && isRetryableError(lastError, config.retryableStatusCodes);
            if (!shouldRetry) {
                // Log final failure
                if (attempt >= config.maxRetries) {
                    console.warn(`Request failed after ${attempt + 1} attempts. Final error: ${lastError.message}`);
                }
                else {
                    // Non-retryable error (e.g., 404, 401)
                    console.error(`Non-retryable error: ${lastError.message}`);
                }
                throw lastError;
            }
            // Log retry event
            if (lastError instanceof BraveApiError && lastError.statusCode === 429) {
                const rateInfo = lastError.body?.error?.meta;
                if (rateInfo?.rate_limit) {
                    console.info(`Rate limited. Current usage: ${rateInfo.rate_current}/${rateInfo.rate_limit} per second. Retrying...`);
                }
                else {
                    console.info(`Rate limited (HTTP 429). Retrying (attempt ${attempt + 1}/${config.maxRetries})...`);
                }
            }
            else {
                console.info(`Request failed (HTTP ${lastError instanceof BraveApiError ? lastError.statusCode : 'Network'}). Retrying (attempt ${attempt + 1}/${config.maxRetries})...`);
            }
        }
    }
    // This should never be reached, but TypeScript needs it
    throw lastError;
}
/**
 * Custom error class for API errors with additional context
 */
export class BraveApiError extends Error {
    statusCode;
    body;
    retryAfter;
    constructor(statusCode, body, retryAfter) {
        super(`HTTP ${statusCode}: ${body?.error?.message || 'Request failed'}`);
        this.statusCode = statusCode;
        this.body = body;
        this.retryAfter = retryAfter;
        this.name = 'BraveApiError';
    }
}
