// @dyingc/brave-search-mcp-server/dist/config.js
import { LoggingLevelSchema } from '@modelcontextprotocol/sdk/types.js';
import { Command } from 'commander';
import dotenv from 'dotenv';
import { z } from 'zod';
import tools from './tools/index.js';
dotenv.config({ debug: false, quiet: true });
// Config schema for Smithery.ai
export const configSchema = z.object({
    braveApiKey: z
        .string()
        .describe('Your API key')
        .default(process.env.BRAVE_API_KEY ?? ''),
    enabledTools: z
        .array(z.string())
        .describe('Enforces a tool whitelist (cannot be used with disabledTools)')
        .optional(),
    disabledTools: z
        .array(z.string())
        .describe('Enforces a tool blacklist (cannot be used with enabledTools)')
        .optional(),
    loggingLevel: z
        .enum([
        'debug',
        'error',
        'info',
        'notice',
        'warning',
        'critical',
        'alert',
        'emergency',
    ])
        .default('info')
        .describe('Desired logging level')
        .optional(),
    stateless: z
        .boolean()
        .default(false)
        .describe('Whether the server should be stateless')
        .optional(),
    apiKeySelectionStrategy: z
        .enum(['greedy-max', 'greedy-min', 'soft-max'])
        .default('soft-max')
        .describe('API key selection strategy')
        .optional(),
    apiKeyTemperature: z
        .number()
        .min(0)
        .max(5.0)
        .default(1.0)
        .describe('Temperature for soft-max strategy (0-5.0, 0 = hard limit)')
        .optional(),
    apiKeyExplorationRate: z
        .number()
        .min(0)
        .max(1.0)
        .default(0.1)
        .describe('Epsilon for epsilon-greedy exploration (0-1, 0 = no exploration, 1 = always random)')
        .optional(),
});
const state = {
    transport: 'stdio',
    port: 8080,
    host: '0.0.0.0',
    braveApiKeys: [],
    loggingLevel: 'info',
    ready: false,
    enabledTools: [],
    disabledTools: [],
    stateless: false,
    retryMaxAttempts: 5,
    retryBaseDelay: 1000,
    retryMaxDelay: 30000,
    apiKeySelectionStrategy: 'soft-max',
    apiKeyTemperature: 1.0,
    apiKeyExplorationRate: 0.1,
};
export function isToolPermittedByUser(toolName) {
    return state.enabledTools.length > 0
        ? state.enabledTools.includes(toolName)
        : state.disabledTools.includes(toolName) === false;
}
/**
 * Parse API keys from multiple sources:
 * 1. CLI argument --brave-api-key (supports comma-separated)
 * 2. CLI argument --brave-api-keys (supports comma-separated)
 * 3. Environment variable BRAVE_API_KEYS (comma-separated)
 * 4. Numbered environment variables BRAVE_API_KEY_1, BRAVE_API_KEY_2, ... (legacy)
 */
function parseApiKeys(options) {
    const keys = [];
    // From CLI argument --brave-api-key (supports comma-separated)
    if (options.braveApiKey) {
        keys.push(...options.braveApiKey.split(',').map((k) => k.trim()));
    }
    // From CLI argument --brave-api-keys (supports comma-separated)
    if (options.braveApiKeys) {
        keys.push(...options.braveApiKeys.split(',').map((k) => k.trim()));
    }
    // From BRAVE_API_KEYS environment variable (comma-separated)
    if (process.env.BRAVE_API_KEYS) {
        keys.push(...process.env.BRAVE_API_KEYS.split(',').map((k) => k.trim()));
    }
    // From BRAVE_API_KEY_1, BRAVE_API_KEY_2, ... (legacy support)
    let i = 1;
    while (process.env[`BRAVE_API_KEY_${i}`]) {
        keys.push(process.env[`BRAVE_API_KEY_${i}`].trim());
        i++;
    }
    // Remove duplicates and empty strings
    return [...new Set(keys)].filter((k) => k.length > 0);
}
export function getOptions() {
    const program = new Command()
        .option('--brave-api-key <string>', 'Brave API key(s) - comma-separated for multiple', process.env.BRAVE_API_KEY ?? '')
        .option('--brave-api-keys <string>', 'Brave API key(s) - comma-separated for multiple', process.env.BRAVE_API_KEYS ?? '')
        .option('--logging-level <string>', 'Logging level', process.env.BRAVE_MCP_LOG_LEVEL ?? 'info')
        .option('--transport <stdio|http>', 'transport type', process.env.BRAVE_MCP_TRANSPORT ?? 'stdio')
        .option('--enabled-tools <names...>', 'tools to enable', process.env.BRAVE_MCP_ENABLED_TOOLS?.trim().split(' ') ?? [])
        .option('--disabled-tools <names...>', 'tools to disable', process.env.BRAVE_MCP_DISABLED_TOOLS?.trim().split(' ') ?? [])
        .option('--port <number>', 'desired port for HTTP transport', process.env.BRAVE_MCP_PORT ?? '8080')
        .option('--host <string>', 'desired host for HTTP transport', process.env.BRAVE_MCP_HOST ?? '0.0.0.0')
        .option('--stateless <boolean>', 'whether the server should be stateless', process.env.BRAVE_MCP_STATELESS === 'true' ? true : false)
        .option('--retry-max-attempts <number>', 'Maximum retry attempts for rate-limited requests', process.env.BRAVE_MCP_MAX_RETRIES ?? '5')
        .option('--retry-base-delay <number>', 'Base delay for retry in milliseconds', process.env.BRAVE_MCP_RETRY_BASE_DELAY ?? '1000')
        .option('--retry-max-delay <number>', 'Maximum delay for retry in milliseconds', process.env.BRAVE_MCP_RETRY_MAX_DELAY ?? '30000')
        .option('--api-key-selection-strategy <strategy>', 'API key selection strategy (greedy-max, greedy-min, soft-max)', process.env.BRAVE_API_KEY_SELECTION_STRATEGY ?? 'soft-max')
        .option('--api-key-temperature <number>', 'Temperature for soft-max strategy (0-5.0, default: 1.0, 0 = hard limit)', process.env.BRAVE_API_KEY_TEMPERATURE ?? '1.0')
        .option('--api-key-exploration-rate <number>', 'Epsilon for epsilon-greedy exploration (0-1, default: 0.1)', process.env.BRAVE_API_KEY_EXPLORATION_RATE ?? '0.1')
        .allowUnknownOption()
        .parse(process.argv);
    const options = program.opts();
    const toolNames = Object.values(tools).map((tool) => tool.name);
    // Validate tool inclusion configuration
    const { enabledTools, disabledTools } = options;
    if (enabledTools.length > 0 && disabledTools.length > 0) {
        console.error('Error: --enabled-tools and --disabled-tools cannot be used together');
        return false;
    }
    if ([...enabledTools, ...disabledTools].some((t) => t.trim().length > 0 && !toolNames.includes(t.trim()))) {
        console.error(`Invalid tool name used. Must be one of: ${toolNames.join(', ')}`);
        return false;
    }
    // Validate all other options
    if (!['stdio', 'http'].includes(options.transport)) {
        console.error(`Invalid --transport value: '${options.transport}'. Must be one of: stdio, http.`);
        return false;
    }
    if (!LoggingLevelSchema.options.includes(options.loggingLevel)) {
        console.error(`Invalid --logging-level value: '${options.loggingLevel}'. Must be one of: ${LoggingLevelSchema.options.join(', ')}`);
        return false;
    }
    // Parse and validate API keys
    const parsedKeys = parseApiKeys(options);
    if (parsedKeys.length === 0) {
        console.error('Error: --brave-api-key, --brave-api-keys, or BRAVE_API_KEYS is required. Get one at https://brave.com/search/api/');
        return false;
    }
    options.braveApiKeys = parsedKeys;
    // Validate retry configuration
    const retryMaxAttempts = parseInt(options.retryMaxAttempts);
    if (isNaN(retryMaxAttempts) || retryMaxAttempts < 0 || retryMaxAttempts > 10) {
        console.error(`Invalid --retry-max-attempts value: '${options.retryMaxAttempts}'. Must be a number between 0 and 10.`);
        return false;
    }
    options.retryMaxAttempts = retryMaxAttempts;
    const retryBaseDelay = parseInt(options.retryBaseDelay);
    if (isNaN(retryBaseDelay) || retryBaseDelay < 100 || retryBaseDelay > 60000) {
        console.error(`Invalid --retry-base-delay value: '${options.retryBaseDelay}'. Must be a number between 100 and 60000.`);
        return false;
    }
    options.retryBaseDelay = retryBaseDelay;
    const retryMaxDelay = parseInt(options.retryMaxDelay);
    if (isNaN(retryMaxDelay) || retryMaxDelay < 1000 || retryMaxDelay > 300000) {
        console.error(`Invalid --retry-max-delay value: '${options.retryMaxDelay}'. Must be a number between 1000 and 300000.`);
        return false;
    }
    options.retryMaxDelay = retryMaxDelay;
    // Validate API key selection strategy
    const validStrategies = ['greedy-max', 'greedy-min', 'soft-max'];
    if (!validStrategies.includes(options.apiKeySelectionStrategy)) {
        console.error(`Invalid --api-key-selection-strategy value: '${options.apiKeySelectionStrategy}'. Must be one of: ${validStrategies.join(', ')}`);
        return false;
    }
    // Validate temperature (range 0-5.0)
    const temperature = parseFloat(options.apiKeyTemperature);
    if (isNaN(temperature) || temperature < 0 || temperature > 5.0) {
        console.error(`Invalid --api-key-temperature value: '${options.apiKeyTemperature}'. Must be a number between 0 and 5.0`);
        return false;
    }
    options.apiKeyTemperature = temperature;
    options.apiKeySelectionStrategy = options.apiKeySelectionStrategy;
    // Validate exploration rate (range 0-1)
    const explorationRate = parseFloat(options.apiKeyExplorationRate ?? '0.1');
    if (isNaN(explorationRate) || explorationRate < 0 || explorationRate > 1) {
        console.error(`Invalid --api-key-exploration-rate value: '${options.apiKeyExplorationRate}'. Must be a number between 0 and 1`);
        return false;
    }
    options.apiKeyExplorationRate = explorationRate;
    if (options.transport === 'http') {
        if (options.port < 1 || options.port > 65535) {
            console.error(`Invalid --port value: '${options.port}'. Must be a valid port number between 1 and 65535.`);
            return false;
        }
        if (!options.host) {
            console.error('Error: --host is required');
            return false;
        }
    }
    // Normalize stateless to boolean (CLI passes it as string)
    options.stateless = options.stateless === true || options.stateless === 'true';
    // Update state
    state.braveApiKeys = options.braveApiKeys;
    state.transport = options.transport;
    state.port = options.port;
    state.host = options.host;
    state.loggingLevel = options.loggingLevel;
    state.enabledTools = options.enabledTools;
    state.disabledTools = options.disabledTools;
    state.stateless = options.stateless;
    state.retryMaxAttempts = options.retryMaxAttempts;
    state.retryBaseDelay = options.retryBaseDelay;
    state.retryMaxDelay = options.retryMaxDelay;
    state.apiKeySelectionStrategy = options.apiKeySelectionStrategy;
    state.apiKeyTemperature = options.apiKeyTemperature;
    state.apiKeyExplorationRate = options.apiKeyExplorationRate;
    state.ready = true;
    return options;
}
export function setOptions(options) {
    return Object.assign(state, options);
}
export default state;
