// @dyingc/brave-search-mcp-server/dist/BraveAPI/index.js
import config from '../config.js';
import { withRetry, BraveApiError } from '../utils/retry.js';
import { extractRetryAfterMs } from './types.js';
import { getApiKeyManager } from '../utils/apiKeyManager.js';
const typeToPathMap = {
    images: '/res/v1/images/search',
    localPois: '/res/v1/local/pois',
    localDescriptions: '/res/v1/local/descriptions',
    news: '/res/v1/news/search',
    videos: '/res/v1/videos/search',
    web: '/res/v1/web/search',
    summarizer: '/res/v1/summarizer/search',
};
const getDefaultRequestHeaders = () => {
    const manager = getApiKeyManager();
    // Use key manager if available, otherwise fallback to first key
    const keyToUse = manager ? manager.selectBestKey() : config.braveApiKeys[0];
    if (!keyToUse) {
        throw new Error('No valid API key available');
    }
    return {
        Accept: 'application/json',
        'Accept-Encoding': 'gzip',
        'X-Subscription-Token': keyToUse,
    };
};
const isValidGoggleURL = (url) => {
    try {
        // Only allow HTTPS URLs
        return new URL(url).protocol === 'https:';
    }
    catch {
        return false;
    }
};
/**
 * Perform the actual HTTP request
 * This function is separated from issueRequest to allow retry logic
 *
 * @param urlWithParams - Full URL with query parameters
 * @param headers - Request headers
 * @returns Promise resolving to the response
 * @throws BraveApiError on HTTP errors
 */
async function performRequest(urlWithParams, headers) {
    const response = await fetch(urlWithParams, { headers });
    // Extract used key for tracking
    const usedApiKey = headers['X-Subscription-Token'];
    const manager = getApiKeyManager();
    // Handle Error
    if (!response.ok) {
        let responseBody;
        try {
            responseBody = await response.json();
        }
        catch {
            // If JSON parsing fails, try to get text
            const text = await response.text();
            responseBody = { message: text };
        }
        // Update key state on error
        if (manager && usedApiKey) {
            if (response.status === 401) {
                manager.markKeyInvalid(usedApiKey, responseBody.message || 'Unauthorized');
            }
            else if (response.status === 429) {
                const retryAfter = extractRetryAfterMs(responseBody);
                manager.markKeyRateLimited(usedApiKey, retryAfter);
            }
            else {
                // Release key for other errors
                manager.releaseKey(usedApiKey);
            }
        }
        // Extract retry-after from rate limit metadata if available
        const retryAfter = extractRetryAfterMs(responseBody);
        // Throw structured error for retry logic to handle
        throw new BraveApiError(response.status, responseBody, retryAfter);
    }
    // Update key state from response headers on success
    if (manager && usedApiKey) {
        manager.updateKeyFromHeaders(usedApiKey, response.headers);
    }
    // Return Response
    const responseBody = await response.json();
    return responseBody;
}
async function issueRequest(endpoint, parameters, 
// TODO (Sampson): Implement support for custom request headers (helpful for POIs, etc.)
requestHeaders = {}) {
    // Determine URL, and setup parameters
    const url = new URL(`https://api.search.brave.com${typeToPathMap[endpoint]}`);
    const queryParams = new URLSearchParams();
    // TODO (Sampson): Move param-construction/validation to modules
    for (const [key, value] of Object.entries(parameters)) {
        // The 'ids' parameter is expected to appear multiple times for multiple IDs
        if (['localPois', 'localDescriptions'].includes(endpoint)) {
            if (key === 'ids') {
                if (Array.isArray(value) && value.length > 0) {
                    value.forEach((id) => queryParams.append(key, id));
                }
                else if (typeof value === 'string') {
                    queryParams.set(key, value);
                }
                continue;
            }
        }
        // Handle `result_filter` parameter
        if (key === 'result_filter') {
            // Handle special behavior of 'summary' parameter:
            // Requires `result_filter` to be empty, or only contain 'summarizer'
            // see: https://bravesoftware.slack.com/archives/C01NNFM9XMM/p1751654841090929
            if ('summary' in parameters && parameters.summary === true) {
                queryParams.set(key, 'summarizer');
            }
            else if (Array.isArray(value) && value.length > 0) {
                queryParams.set(key, value.join(','));
            }
            continue;
        }
        // Handle `goggles` parameter(s)
        if (key === 'goggles') {
            if (typeof value === 'string') {
                queryParams.set(key, value);
            }
            else if (Array.isArray(value)) {
                for (const url of value.filter(isValidGoggleURL)) {
                    queryParams.append(key, url);
                }
            }
            continue;
        }
        if (value !== undefined) {
            queryParams.set(key === 'query' ? 'q' : key, value.toString());
        }
    }
    // Build final URL and headers
    const urlWithParams = url.toString() + '?' + queryParams.toString();
    const headers = { ...getDefaultRequestHeaders(), ...requestHeaders };
    // Special case: Summarizer uses its own polling logic, don't apply retry
    if (endpoint === 'summarizer') {
        return performRequest(urlWithParams, headers);
    }
    // All other endpoints: Use retry logic with exponential backoff
    return withRetry(() => performRequest(urlWithParams, headers), {
        maxRetries: config.retryMaxAttempts,
        baseDelay: config.retryBaseDelay,
        maxDelay: config.retryMaxDelay,
        onRetry: (attempt, error, delay) => {
            // Additional logging can be added here if needed
            if (error instanceof BraveApiError && error.statusCode === 429) {
                // Log specific rate limit info if available
                if (error.retryAfter) {
                    console.info(`Rate limit metadata suggests ${error.retryAfter}ms delay`);
                }
            }
        },
    });
}
export default {
    issueRequest,
};
